assembly/src/tests.rs

```diff
@@ -105,6 +108,71 @@ fn span_and_simple_if() {
     assert_eq!(expected, format!("{program}"));
 }
 
+// PROGRAM WITH #main CALL
+// ================================================================================================
+#[test]
+fn simple_main_call() {
+    // instantiate assembler
+    let assembler = super::Assembler::default();
+
+    // compile account module
+    let account_path = LibraryPath::new("context::account").unwrap();
+    let account_code = ModuleAst::parse(
+        "\
+    export.account_method_1
+        push.2.1 add
+    end
+    
+    export.account_method_2
+        push.3.1 sub
+    end
+    ",
+    )
+    .unwrap();
+    let account_module = Module::new(account_path, account_code);
+    let _method_roots = assembler
+        .compile_module(&account_module, &mut super::AssemblyContext::default())
+        .unwrap();
+
+    // define note main body procedure name
+    let procedure_name: ProcedureName = String::from("main").try_into().unwrap();
+
+    // compile note 1 program
+    let mut assembler_context = super::AssemblyContext::new_program(
+        LibraryPath::new("context::note1").unwrap(),
+        procedure_name.clone(),
+    );
+    let note_1 =
+        ProgramAst::parse("use.context::account begin exec.account::account_method_1 end").unwrap();
+    let _note_1_root = assembler.compile_in_context(note_1, &mut assembler_context).unwrap();
+
+    // compile note 2 program
+    let mut assembler_context = super::AssemblyContext::new_program(
+        LibraryPath::new("context::note2").unwrap(),
+        procedure_name,
+    );
+    let note_2 =
+        ProgramAst::parse("use.context::account begin exec.account::account_method_2 end").unwrap();
+    let _note_2_root = assembler.compile_in_context(note_2, &mut assembler_context).unwrap();
+
+    // compile transaction program
+    let source = "\
+    use.context::note1 
+    use.context::note2
+    
+    # these calls will need to be modified to use MAST roots
+    begin call.note1::main call.note2::main end";
```

Comment:
I'm actually not sure this is the right way to build transaction programs. Two reasons for this:

First, transaction programs need to include prologue, epilogue, note setup script etc. So, the way this should look like would be:
```
use.miden::sat::internal::prologue
use.miden::sat::internal::epilogue
use.miden::sat::internal::note_setup
use.context::note1 
use.context::note2

begin
  exec.prologue::prepare_transaction

  exec.note_setup::prepare_note
  call.note1::main
  
  exec.note_setup::prepare_note
  call.note2::main

  exec.epilogue::finalize_transaction
end
```

Second, I'm not sure we actually ever need to deal with textual representation of the above. The way I'm imagining we'd build a transaction program is not by writing the above MASM and compiling it, but by combining blocks manually. Specifically:

- We'd have a pre-build `CodeBlock`'s for prologue, epilogue, and note setup script (together with their associated callsets).
- For each transaction, we'd build `CodeBlock`'s for each note using `compile_in_context()`.
- And then we'd "manually" join these blocks together using constructors of `new_join()` and `new_call()` constructors of `CodeBlock` (we'd also need to merge their respective callsets).

To perform the above procedure, we could have a component in `miden-base` (or somewhere else) - maybe something like `TxCompiler` which would internally maintain pre-built `CodeBlock`'s for prologue, epilogue etc. as well as an instance of the assembler.

Btw, with the introduction of `pcall` much of the above complexity would go away as we'd have a single program for transaction execution which could look something like:
```
use.miden::sat::internal::prologue
use.miden::sat::internal::epilogue
use.miden::sat::internal::note_setup

begin
  exec.prologue::prepare_transaction

  while.true
    exec.note_setup::prepare_note
    pcall
  end

  exec.epilogue::finalize_transaction
end
```

Reply:
> First, transaction programs need to include prologue, epilogue, note setup script etc. So, the way this should look like would be:
> 
> ```
> use.miden::sat::internal::prologue
> use.miden::sat::internal::epilogue
> use.miden::sat::internal::note_setup
> use.context::note1 
> use.context::note2
> 
> begin
>   exec.prologue::prepare_transaction
> 
>   exec.note_setup::prepare_note
>   call.note1::main
>   
>   exec.note_setup::prepare_note
>   call.note2::main
> 
>   exec.epilogue::finalize_transaction
> end
> ```

yep, I left out the other components for brevity. 

> Second, I'm not sure we actually ever need to deal with textual representation of the above. The way I'm imagining we'd build a transaction program is not by writing the above MASM and compiling it, but by combining blocks manually. Specifically:
> 
> * We'd have a pre-build `CodeBlock`'s for prologue, epilogue, and note setup script (together with their associated callsets).

Having pre-compiles for this makes a lot of sense, I like this idea!

One way we could support this in the `Assembler` is to introduce a `pre_compiles` component:

```rust
pub struct Assembler<'a> {
    kernel: Kernel,
    module_provider: ModuleProvider,
    proc_cache: RefCell<ProcedureCache>,
    in_debug_mode: bool,
    pre_compiles: Option<&'a ProcedureCache>,
}
```

> * For each transaction, we'd build `CodeBlock`'s for each note using `compile_in_context()`.

I'm guessing we would then need to also return the `CodeBlockTable` for each note such we have access to the call target procedures?  In essence we would return a `Program` struct?  If we only return the `CallSet` then I'm not sure how the client will have access to the associated procedures?

> * And then we'd "manually" join these blocks together using constructors of `new_join()` and `new_call()` constructors of `CodeBlock` (we'd also need to merge their respective callsets).
> 
> To perform the above procedure, we could have a component in `miden-base` (or somewhere else) - maybe something like `TxCompiler` which would internally maintain pre-built `CodeBlock`'s for prologue, epilogue etc. as well as an instance of the assembler.

It seems to me that the `TxCompiler` would be replicating a lot of logic of the `Assembler`.  On the surface I think it would be easier to extend the `Assembler` to support this compilation pattern then to create a `TxCompiler` object.    

> Btw, with the introduction of `pcall` much of the above complexity would go away as we'd have a single program for transaction execution which could look something like:
> 
> ```
> use.miden::sat::internal::prologue
> use.miden::sat::internal::epilogue
> use.miden::sat::internal::note_setup
> 
> begin
>   exec.prologue::prepare_transaction
> 
>   while.true
>     exec.note_setup::prepare_note
>     pcall
>   end
> 
>   exec.epilogue::finalize_transaction
> end
> ```

Agreed!  `pcall` will make things easier, we will just need to specify call targets such that the `CodeBlockTable` can be constructed appropriately.